### PR TITLE
Orders logic

### PR DESCRIPTION
**Pull request by artemklishch**
*Made order controller, order/order item services, repositories,, dtos, mappers
*Made enum validator
*Made enums package and moved enums to there

Task: https://mate.academy/learn/java-spring-boot/java-spring-boot-practice-v2/linear?learnItemsFilter=All&section=practice&testTaskSlug=java_add_order_model&taskSection=task-description

### CODE REVIEW

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                );
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        return getCart(authentication);
+    }
+
+    @Override
+    public ShoppingCartDto deleteCartItem(Long cartItemId, Authentication authentication) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
```
Comment (okuzan) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: Deal with authentication on controller level, extract repeated logic to a private method, service shouldn't be bothered with this

File: src/main/resources/db/changelog/changes/18-create-order-items-table.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+              - column:
+                  name: quantity
+                  type: int
+                  constraints:
+                    nullable: false
+              - column:
+                  name: price
+                  type: decimal
+                  constraints:
+                    nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/18-create-order-items-table.yaml]:
```suggestion
                    nullable: false

```

File: src/main/java/org/example/intro/validation/EnumValidator.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        if (value == null) {
+            return true;
+        }
+
+        return Arrays.stream(enumClass.getEnumConstants())
+                .anyMatch(enumConstant -> enumConstant.name().equals(value));
+    }
+}
```
Comment (okuzan) [inline src/main/java/org/example/intro/validation/EnumValidator.java]:
```suggestion
}

```

File: src/main/java/org/example/intro/validation/EnumValidator.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package org.example.intro.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+import java.util.Arrays;
+
+public class EnumValidator implements ConstraintValidator<EnumMatch, String> {
+
```
Comment (okuzan) [inline src/main/java/org/example/intro/validation/EnumValidator.java]:
```suggestion
```

File: src/main/java/org/example/intro/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        );
+    }
+
+    @Override
+    public OrderDto updateStatus(Long orderId, UpdateOrderStatusDto requestDto) {
+        Order order = orderRepository.findById(orderId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Not found the order with the id: " + orderId
+                )
+        );
```
Comment (okuzan) [inline src/main/java/org/example/intro/service/impl/OrderServiceImpl.java]: extract to a private method, you use it at least twice

File: src/main/java/org/example/intro/service/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public List<OrderItemDto> getOrderItems(
+            Long orderId, Authentication authentication
+    ) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        return orderRepository.findOrderItems(orderId, userId)
+                .stream()
+                .map(orderItemMapper::toDto)
+                .toList();
```
Comment (okuzan) [inline src/main/java/org/example/intro/service/impl/OrderServiceImpl.java]: not a mistake but fyi we can declare mapping list to list in mapstruct interface

File: src/main/java/org/example/intro/repository/order/OrderRepository.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    @EntityGraph(attributePaths = "orderItems.book")
+    List<Order> findByUserId(Long userId, Pageable pageable);
+
+    @Query("SELECT o.orderItems FROM Order o WHERE o.id=?1 AND o.user.id=?2")
+    List<OrderItem> findOrderItems(Long orderId, Long userId);
+
+    @Query("SELECT o FROM Order o WHERE o.id=?1 AND o.user.id=?2")
+    Optional<Order> findOrder(Long orderId, Long userId);
+
```
Comment (okuzan) [inline src/main/java/org/example/intro/repository/order/OrderRepository.java:21]:
```suggestion
```

File: src/main/java/org/example/intro/repository/order/OrderRepository.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    @EntityGraph(attributePaths = "orderItems.book")
+    List<Order> findByUserId(Long userId, Pageable pageable);
+
+    @Query("SELECT o.orderItems FROM Order o WHERE o.id=?1 AND o.user.id=?2")
```
Comment (okuzan) [inline src/main/java/org/example/intro/repository/order/OrderRepository.java]:
```suggestion
    @Query("SELECT o.orderItems FROM Order o WHERE o.id= ?1 AND o.user.id= ?2")
```

File: src/main/java/org/example/intro/repository/order/OrderItemRepository.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package org.example.intro.repository.order;
+
+import org.example.intro.model.OrderItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+    @Query("SELECT i FROM OrderItem i WHERE i.order.id=?1 AND i.id=?2")
```
Comment (okuzan) [inline src/main/java/org/example/intro/repository/order/OrderItemRepository.java:8]:
```suggestion
    @Query("SELECT i FROM OrderItem i WHERE i.order.id = ?1 AND i.id = ?2")
```

File: src/main/java/org/example/intro/repository/cart/CartItemRepository.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package org.example.intro.repository.cart;
+
+import org.example.intro.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Query("SELECT c FROM CartItem c WHERE c.shoppingCart.id= ?1 AND c.book.id= ?2")
```
Comment (okuzan) [inline src/main/java/org/example/intro/repository/cart/CartItemRepository.java:8]:
```suggestion
    @Query("SELECT c FROM CartItem c WHERE c.shoppingCart.id = ?1 AND c.book.id = ?2")
```

File: src/main/java/org/example/intro/repository/cart/CartItemRepository.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    CartItem findByCartIdAndBookId(Long cartId, Long bookId);
+
+    @Query("SELECT c FROM CartItem c WHERE c.shoppingCart.id= ?1 AND c.id= ?2")
```
Comment (okuzan) [inline src/main/java/org/example/intro/repository/cart/CartItemRepository.java:11]:
```suggestion
    @Query("SELECT c FROM CartItem c WHERE c.shoppingCart.id = ?1 AND c.id = ?2")
```

File: src/main/java/org/example/intro/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted=false")
```
Comment (okuzan) [inline src/main/java/org/example/intro/model/ShoppingCart.java]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction("is_deleted = FALSE")
```

File: src/main/java/org/example/intro/dto/cart/CartItemDto.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.Setter;
+
+@Setter
+@Getter
+public class CartItemDto{
+    private Long id;
+    private Long bookId;
+    private String bookTitle;
+    private int quantity;
+}
```
Comment (okuzan) [inline src/main/java/org/example/intro/dto/cart/CartItemDto.java]:
```suggestion
}

```

File: src/main/java/org/example/intro/dto/cart/CartItemDto.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package org.example.intro.dto.cart;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Setter
+@Getter
```
Comment (okuzan) [inline src/main/java/org/example/intro/dto/cart/CartItemDto.java]: not a mistake - it's ok to just use @Data on Dtos, nothing dangerous here (unlike entities)

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job overall! mostly codestyle suggestions

File: src/main/java/org/example/intro/dto/cart/CartItemDto.java (hunk 14 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/cart/CartItemDto.java]: ++, updated. thank you.

File: src/main/java/org/example/intro/dto/cart/CartItemDto.java (hunk 13 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/cart/CartItemDto.java]: ++

File: src/main/java/org/example/intro/model/ShoppingCart.java (hunk 12 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/ShoppingCart.java]: ++

File: src/main/java/org/example/intro/repository/cart/CartItemRepository.java (hunk 11 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/cart/CartItemRepository.java:11]: ++

File: src/main/java/org/example/intro/repository/cart/CartItemRepository.java (hunk 10 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/cart/CartItemRepository.java:8]: ++

File: src/main/java/org/example/intro/repository/order/OrderItemRepository.java (hunk 9 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/order/OrderItemRepository.java:8]: ++

File: src/main/java/org/example/intro/repository/order/OrderRepository.java (hunk 8 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/order/OrderRepository.java]: ++

File: src/main/java/org/example/intro/repository/order/OrderRepository.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/repository/order/OrderRepository.java:21]: ++

File: src/main/java/org/example/intro/service/impl/OrderServiceImpl.java (hunk 6 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/OrderServiceImpl.java]: ++, updated. thank you!

File: src/main/java/org/example/intro/service/impl/OrderServiceImpl.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/OrderServiceImpl.java]: ++, made additional repository method for that

File: src/main/java/org/example/intro/validation/EnumValidator.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/validation/EnumValidator.java]: ++

File: src/main/java/org/example/intro/validation/EnumValidator.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/validation/EnumValidator.java]: ++

File: src/main/resources/db/changelog/changes/18-create-order-items-table.yaml (hunk 2 above)
Comment (artemklishch) [reply src/main/resources/db/changelog/changes/18-create-order-items-table.yaml]: ++, removed

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 1 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++, updated. Thank you!

**Review by sarakhmen — APPROVED** [review]
lgtm! Do not forget to merge your changes :)

**Pull request merged**